src/store/indexeddb.js

```diff
@@ -0,0 +1,403 @@
+                return user;
+            });
+        });
+    },
+
+    /**
+     * Load all the account data events from the database. This is not cached.
+     * @return {Promise<MatrixEvent[]>} A list of events.
+     */
+    loadAccountData: function() {
+        return q.try(() => {
+            const txn = this.db.transaction(["accountData"], "readonly");
+            const store = txn.objectStore("accountData");
+            return selectQuery(store, undefined, (cursor) => {
+                return new MatrixEvent(cursor.value);
+            });
+        });
+    },
+
+    /**
+     * Load the sync data from the database.
+     * @return {Promise<Object>} An object with "roomsData" and "nextBatch" keys.
+     */
+    loadSyncData: function() {
+        return q.try(() => {
+            const txn = this.db.transaction(["sync"], "readonly");
+            const store = txn.objectStore("sync");
+            return selectQuery(store, undefined, (cursor) => {
+                return cursor.value;
+            }).then((results) => {
+                if (results.length > 1) {
+                    console.warn("loadSyncData: More than 1 sync row found.");
+                }
+                return (results.length > 0 ? results[0] : {});
+            });
+        });
+    },
+};
+
+/**
+ * Construct a new Indexed Database store, which extends MatrixInMemoryStore.
+ *
+ * This store functions like a MatrixInMemoryStore except it periodically persists
+ * the contents of the store to an IndexedDB backend.
+ *
+ * All data is still kept in-memory but can be loaded from disk by calling
+ * <code>startup()</code>. This can make startup times quicker as a complete
+ * sync from the server is not required. This does not reduce memory usage as all
+ * the data is eagerly fetched when <code>startup()</code> is called.
+ * <pre>
+ * let syncAccumulator = new SyncAccumulator();
+ * let opts = { localStorage: window.localStorage };
+ * let store = new IndexedDBStore(
+ *     new IndexedDBStoreBackend(window.indexedDB), syncAccumulator, opts
+ * );
+ * await store.startup(); // load from indexed db
+ * let client = sdk.createClient({
+ *     store: store,
+ * });
+ * client.startClient();
+ * client.on("sync", function(state, prevState, data) {
+ *     if (state === "PREPARED") {
+ *         console.log("Started up, now with go faster stripes!");
+ *     }
+ * });
+ * </pre>
+ *
+ * @constructor
+ * @extends MatrixInMemoryStore
+ * @param {IndexedDBStoreBackend} backend The indexed db backend instance.
+ * @param {SyncAccumulator} syncAccumulator The sync accumulator which will be
+ * loaded from IndexedDB and periodically saved to IndexedDB.
+ * @param {Object=} opts Options for MatrixInMemoryStore.
+ * @prop {IndexedDBStoreBackend} backend The backend instance. Call through to
+ * this API if you need to perform specific indexeddb actions like deleting the
+ * database.
+ */
+const IndexedDBStore = function IndexedDBStore(backend, syncAccumulator, opts) {
```

Comment:
We do this because of JSDoc: https://github.com/matrix-org/matrix-js-sdk/pull/363#discussion_r101494996